python/ql/src/semmle/python/dataflow/Implementation.qll

```diff
@@ -0,0 +1,849 @@
+            node.asCfgNode() = call and
+            retval.asCfgNode() = any(Return ret | ret.getScope() = pyfunc.getScope()).getValue().getAFlowNode()
+        ) and
+        edgeLabel = "return"
+    }
+
+    /* Avoid taint flow from return value to caller as it can produce imprecise flow graphs
+     * Step directly from tainted argument to call result.
+     */
+    pragma [noinline]
+    predicate callFlowStep(TaintTrackingNode src, DataFlow::Node node, TaintTrackingContext context, AttributePath path, TaintKind kind, string edgeLabel) {
+        exists(CallNode call, PythonFunctionObjectInternal pyfunc, TaintTrackingContext callee, DataFlow::Node retval, TaintTrackingNode retnode |
+            this.callContexts(call, src, pyfunc, context, callee) and
+            retnode = TTaintTrackingNode_(retval, callee, path, kind, this) and
+            node.asCfgNode() = call and
+            retval.asCfgNode() = any(Return ret | ret.getScope() = pyfunc.getScope()).getValue().getAFlowNode()
+        ) and
+        edgeLabel = "call"
+    }
+
+    pragma [noinline]
+    predicate callContexts(CallNode call, TaintTrackingNode argnode, PythonFunctionObjectInternal pyfunc, TaintTrackingContext caller, TaintTrackingContext callee) {
+        exists(int arg, TaintKind callerKind, AttributePath callerPath |
+            this.callWithTaintedArgument(argnode, call, caller, pyfunc, arg, callerPath, callerKind) and
+            callee = TParamContext(callerKind, callerPath, arg)
+        )
+    }
+
+    predicate callWithTaintedArgument(TaintTrackingNode src, CallNode call, TaintTrackingContext caller, CallableValue pyfunc, int arg, AttributePath path, TaintKind kind) {
+        exists(DataFlow::Node srcnode |
+            src = TTaintTrackingNode_(srcnode, caller, path, kind, this) and
+            srcnode.asCfgNode() = pyfunc.getArgumentForCall(call, arg)
+        )
+    }
+
+    predicate instantiationStep(TaintTrackingNode src, DataFlow::Node node, TaintTrackingContext context, AttributePath path, TaintKind kind, string edgeLabel) {
+        exists(PythonFunctionValue init, EssaVariable self, TaintTrackingContext callee |
+            instantiationCall(node.asCfgNode(), src, init, context, callee) and
+            this.(EssaTaintTracking).taintedDefinition(_, self.getDefinition(), callee, path, kind) and
+            self.getSourceVariable().(Variable).isSelf() and
+            BaseFlow::reaches_exit(self) and
+            self.getScope() = init.getScope()
+        ) and
+        edgeLabel = "instantiation"
+    }
+
+    predicate instantiationCall(CallNode call, TaintTrackingNode argnode, PythonFunctionObjectInternal init, TaintTrackingContext caller, TaintTrackingContext callee) {
+        exists(ClassValue cls |
+            call.getFunction().pointsTo(cls) and
+            cls.lookup("__init__") = init
+            |
+            exists(int arg, TaintKind callerKind, AttributePath callerPath |
+                exists(DataFlow::Node argument |
```

Comment:
Any reason not to merge these two `exists`?

Reply:
done